### PR TITLE
#281 DISABLE System Package cache by default

### LINKED ISSUE #281: DE not finding packages in WEB-INF/lib
While upgrading an Alfresco 4.2.f container with DE 1.7.6 to Alfresco 5.2.g with DE 2.0.2, DE bundles are not started because of missing requirements even though I can find these missing packages in the Alfresco container's WEB-INF/lib.

## I'm submitting a ...   (check one with "x")
```
[X] bug report
[ ] feature request
[ ] question
```

## Expected Behavior

After upgrading the DE module, DE bundles are started and the control panel is reachable.

## Current Behavior

DE module is upgraded, but then throws missing requirement errors when trying to start bundles. Alfresco does start up fine after this.
I got the following error when upgrading from an Alfresco 4.2.f with DE 1.7.6 to Alfresco 5.2.g with only DE 2.0.2 added:

```
 [ALFRESCO] 2020-01-15 15:29:55,648  INFO  [repo.module.ModuleServiceImpl] [localhost-startStop-1] Upgrading module 'alfresco-dynamic-extensions-repo' version 2.0.2 (was 1.7.6).
 [ALFRESCO] 2020-01-15 15:29:55,760  WARN  [context.support.PropertySourcesPlaceholderConfigurer] [localhost-startStop-1] Could not load properties from class path resource [dynamic-extensions/osgi-container.properties]: class path resource [dynamic-extensions/osgi-container.properties] cannot be opened because it does not exist
 [ALFRESCO] 2020-01-15 15:29:56,227  WARN  [dynamicextensionsalfresco.osgi.DefaultFrameworkManager] [localhost-startStop-1] Could not resolve all 8 bundles.
 [ALFRESCO] 2020-01-15 15:29:56,415  ERROR [dynamicextensionsalfresco.osgi.DefaultFrameworkManager] [localhost-startStop-1] Bundle eu.xenit.de.blueprint-integration failed to resolve.
 [ALFRESCO] 2020-01-15 15:29:56,415  ERROR [dynamicextensionsalfresco.osgi.DefaultFrameworkManager] [localhost-startStop-1] Bundle eu.xenit.de.control-panel failed to resolve.
 [ALFRESCO] 2020-01-15 15:29:56,418  ERROR [dynamicextensionsalfresco.osgi.DefaultFrameworkManager] [localhost-startStop-1] Error starting bundle eu.xenit.de.control-panel:
 org.osgi.framework.BundleException: Unable to resolve eu.xenit.de.control-panel [8](R 8.0): missing requirement [eu.xenit.de.control-panel [8](R 8.0)] osgi.wiring.package; (&(osgi.wiring.package=org.slf4j)(version>=1.7.0)(!(version>=2.0.0))) Unresolved requirements: [[eu.xenit.de.control-panel [8](R 8.0)] osgi.wiring.package; (&(osgi.wiring.package=org.slf4j)(version>=1.7.0)(!(version>=2.0.0)))]
	at org.apache.felix.framework.Felix.resolveBundleRevision(Felix.java:4368)
	at org.apache.felix.framework.Felix.startBundle(Felix.java:2281)
	at org.apache.felix.framework.BundleImpl.start(BundleImpl.java:998)
	at org.apache.felix.framework.BundleImpl.start(BundleImpl.java:984)
	at com.github.dynamicextensionsalfresco.osgi.DefaultFrameworkManager.startBundle(DefaultFrameworkManager.java:209)
	at com.github.dynamicextensionsalfresco.osgi.DefaultFrameworkManager.startBundles(DefaultFrameworkManager.java:201)
	at com.github.dynamicextensionsalfresco.osgi.DefaultFrameworkManager.initialize(DefaultFrameworkManager.java:93)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.invokeCustomInitMethod(AbstractAutowireCapableBeanFactory.java:1640)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.invokeInitMethods(AbstractAutowireCapableBeanFactory.java:1581)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.initializeBean(AbstractAutowireCapableBeanFactory.java:1511)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.doCreateBean(AbstractAutowireCapableBeanFactory.java:521)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBean(AbstractAutowireCapableBeanFactory.java:458)
	at org.springframework.beans.factory.support.AbstractBeanFactory$1.getObject(AbstractBeanFactory.java:293)
	at org.springframework.beans.factory.support.DefaultSingletonBeanRegistry.getSingleton(DefaultSingletonBeanRegistry.java:223)
	at org.springframework.beans.factory.support.AbstractBeanFactory.doGetBean(AbstractBeanFactory.java:290)
	at org.springframework.beans.factory.support.AbstractBeanFactory.getBean(AbstractBeanFactory.java:191)
	at org.springframework.beans.factory.support.DefaultListableBeanFactory.preInstantiateSingletons(DefaultListableBeanFactory.java:636)
	at org.springframework.context.support.AbstractApplicationContext.finishBeanFactoryInitialization(AbstractApplicationContext.java:938)
	at org.springframework.context.support.AbstractApplicationContext.refresh(AbstractApplicationContext.java:479)
	at com.github.dynamicextensionsalfresco.osgi.OsgiContainerModuleComponent.initializeOsgiContainerApplicationContext(OsgiContainerModuleComponent.java:109)
	at com.github.dynamicextensionsalfresco.osgi.OsgiContainerModuleComponent.startFramework(OsgiContainerModuleComponent.java:96)
	at com.github.dynamicextensionsalfresco.osgi.OsgiContainerModuleComponent.executeInternal(OsgiContainerModuleComponent.java:56)
	at org.alfresco.repo.module.AbstractModuleComponent.execute(AbstractModuleComponent.java:361)
	at org.alfresco.repo.module.ModuleComponentHelper.executeComponent(ModuleComponentHelper.java:736)
	at org.alfresco.repo.module.ModuleComponentHelper.startModule(ModuleComponentHelper.java:647)
	at org.alfresco.repo.module.ModuleComponentHelper.access$5(ModuleComponentHelper.java:530)
	at org.alfresco.repo.module.ModuleComponentHelper$1$1.execute(ModuleComponentHelper.java:263)
	at org.alfresco.repo.transaction.RetryingTransactionHelper.doInTransaction(RetryingTransactionHelper.java:464)
	at org.alfresco.repo.transaction.RetryingTransactionHelper.doInTransaction(RetryingTransactionHelper.java:352)
	at org.alfresco.repo.module.ModuleComponentHelper$1.doWork(ModuleComponentHelper.java:284)
	at org.alfresco.repo.security.authentication.AuthenticationUtil.runAs(AuthenticationUtil.java:555)
	at org.alfresco.repo.module.ModuleComponentHelper.startModules(ModuleComponentHelper.java:213)
	at org.alfresco.repo.module.ModuleServiceImpl.startModules(ModuleServiceImpl.java:149)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.springframework.aop.support.AopUtils.invokeJoinpointUsingReflection(AopUtils.java:317)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.invokeJoinpoint(ReflectiveMethodInvocation.java:183)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:150)
	at org.springframework.transaction.interceptor.TransactionInterceptor$1.proceedWithInvocation(TransactionInterceptor.java:96)
	at org.springframework.transaction.interceptor.TransactionAspectSupport.invokeWithinTransaction(TransactionAspectSupport.java:260)
	at org.springframework.transaction.interceptor.TransactionInterceptor.invoke(TransactionInterceptor.java:94)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:172)
	at org.springframework.aop.framework.JdkDynamicAopProxy.invoke(JdkDynamicAopProxy.java:204)
	at com.sun.proxy.$Proxy123.startModules(Unknown Source)
	at org.alfresco.repo.module.ModuleStarter$1.execute(ModuleStarter.java:73)
	at org.alfresco.repo.transaction.RetryingTransactionHelper.doInTransaction(RetryingTransactionHelper.java:464)
	at org.alfresco.repo.transaction.RetryingTransactionHelper.doInTransaction(RetryingTransactionHelper.java:352)
	at org.alfresco.repo.module.ModuleStarter$2.doWork(ModuleStarter.java:83)
	at org.alfresco.repo.security.authentication.AuthenticationUtil.runAs(AuthenticationUtil.java:555)
	at org.alfresco.repo.module.ModuleStarter.onBootstrap(ModuleStarter.java:78)
	at org.springframework.extensions.surf.util.AbstractLifecycleBean.onApplicationEvent(AbstractLifecycleBean.java:56)
	at org.alfresco.repo.management.SafeApplicationEventMulticaster.multicastEventInternal(SafeApplicationEventMulticaster.java:214)
	at org.alfresco.repo.management.SafeApplicationEventMulticaster.multicastEvent(SafeApplicationEventMulticaster.java:185)
	at org.springframework.context.support.AbstractApplicationContext.publishEvent(AbstractApplicationContext.java:334)
	at org.springframework.context.support.AbstractApplicationContext.finishRefresh(AbstractApplicationContext.java:954)
	at org.springframework.context.support.AbstractApplicationContext.refresh(AbstractApplicationContext.java:482)
	at org.springframework.web.context.ContextLoader.configureAndRefreshWebApplicationContext(ContextLoader.java:410)
	at org.springframework.web.context.ContextLoader.initWebApplicationContext(ContextLoader.java:306)
	at org.springframework.web.context.ContextLoaderListener.contextInitialized(ContextLoaderListener.java:112)
	at org.alfresco.web.app.ContextLoaderListener.contextInitialized(ContextLoaderListener.java:70)
	at org.apache.catalina.core.StandardContext.listenerStart(StandardContext.java:5197)
	at org.apache.catalina.core.StandardContext.startInternal(StandardContext.java:5720)
	at org.apache.catalina.util.LifecycleBase.start(LifecycleBase.java:145)
	at org.apache.catalina.core.ContainerBase.addChildInternal(ContainerBase.java:1018)
	at org.apache.catalina.core.ContainerBase.addChild(ContainerBase.java:994)
	at org.apache.catalina.core.StandardHost.addChild(StandardHost.java:652)
	at org.apache.catalina.startup.HostConfig.deployDirectory(HostConfig.java:1296)
	at org.apache.catalina.startup.HostConfig$DeployDirectory.run(HostConfig.java:2038)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
```

When I check WEB-INF/lib, it contains slf4j-api-1.7.21.jar and slf4j-log4j12-1.7.21.jar.

## Possible Solution

As a workaround, mounting an osgi-container.properties with osgi.container.system-package-cache.mode set to DISABLE or UPDATE seems to let the bundles start as expected.

## Steps to Reproduce (for bugs)

1. Have a running Alfresco 4.2.f container with DE 1.7.6
2. Upgrade to Alfresco 5.2.g container with DE 2.0.2

Note: I could not reproduce this problem every time, though there's no difference in the steps I take.

## Your Environment

* Alfresco version used: 4.2.f and 5.2.g
* DE version used: 1.7.6 and 2.0.2

### PR DESCRIPTION
**Pull request by kerkhofsd**
Fixes #281

### CODE REVIEW

File: alfresco-integration/src/main/java/com/github/dynamicextensionsalfresco/osgi/spring/WebApplicationSystemPackageFactoryBean.java [hunk 1]
```diff
@@ -47,8 +53,16 @@ public boolean isSingleton() {
 	}
 
 	@Override
-	public Set<SystemPackage> getObject() throws Exception {
-		return createSystemPackages();
+	public Set<SystemPackage> getObject() {
+		final long start = System.currentTimeMillis();
+		Set<SystemPackage> systemPackages = createSystemPackages();
+		final long finish = System.currentTimeMillis();
+		if (logger.isInfoEnabled() && (finish - start) > 1500L) {
```
Comment (vierbergenlars) [inline alfresco-integration/src/main/java/com/github/dynamicextensionsalfresco/osgi/spring/WebApplicationSystemPackageFactoryBean.java]: There is no need to wrap this check into `logger.isInfoEnabled()`.
In general, checking if a log level is enabled is only used when doing expensive operations to create log messages.

File: alfresco-integration/src/main/java/com/github/dynamicextensionsalfresco/osgi/spring/WebApplicationSystemPackageFactoryBean.java [hunk 2]
```diff
@@ -47,8 +53,16 @@ public boolean isSingleton() {
...
+			logger.info("System Package scanning took '{}' milliseconds. Possible improvement: enable System Package "
+					+ "cache by adding 'osgi.container.system-package-cache.mode:ENABLE' "
+					+ "to 'osgi-container.properties' (DISABLE by default because of #281)", (finish - start));
```
Comment (vierbergenlars) [inline alfresco-integration/src/main/java/com/github/dynamicextensionsalfresco/osgi/spring/WebApplicationSystemPackageFactoryBean.java]: Do not refer to an issue number here, it is irrelevant.
You can add a comment referencing the github issue (full url in that case)

File: alfresco-integration/src/main/java/com/github/dynamicextensionsalfresco/osgi/JavaPackageScanner.java [hunk 3]
```diff
@@ -98,7 +98,7 @@
 		} catch (final IOException e) {
 			if (logger.isWarnEnabled()) {
 				logger.warn("Error scanning Java packages. This will cause Dynamic Extensions framework not to work!",
-						e.getMessage(), e);
+						e);
```
Comment (vierbergenlars) [inline alfresco-integration/src/main/java/com/github/dynamicextensionsalfresco/osgi/JavaPackageScanner.java]: Nit: Can you put this on the previous line with the rest of the statement?

Comment (kerkhofsd) [reply alfresco-integration/src/main/java/com/github/dynamicextensionsalfresco/osgi/JavaPackageScanner.java]: Line is too long then...

**Pull request merged**